ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Peras/Voting/Rules.hs

```diff
@@ -0,0 +1,537 @@
+-- within the same STM transaction, or the results may be inconsistent.
+mkPerasVotingView ::
+  ( cert ~ WithArrivalTime (ValidatedPerasCert blk)
+  , GetHeader blk
+  ) =>
+  -- | Peras protocol parameters
+  PerasParams ->
+  -- | Current Peras round number
+  PerasRoundNo ->
+  -- | Most recent certificate seen by the voter
+  WithOrigin cert ->
+  -- | Most recent certificate included in some block in our preferred chain
+  WithOrigin cert ->
+  -- | Prefix of the preferred chain leading to the candidate block
+  AnchoredFragment (Header blk) ->
+  -- | Constructed voting view
+  PerasQry xs (PerasVotingView cert)
+mkPerasVotingView
+  perasParams
+  currRoundNo
+  latestCertSeen
+  latestCertOnChain
+  chainAtCandidateBlock = do
+    latestCertSeenView <- traverse mkLatestCertSeenView latestCertSeen
+    latestCertOnChainView <- traverse mkLatestCertOnChainView latestCertOnChain
+    pure $
+      PerasVotingView
+        { perasParams = perasParams
+        , currRoundNo = currRoundNo
+        , latestCertSeen = latestCertSeenView
+        , latestCertOnChain = latestCertOnChainView
+        }
+   where
+    mkLatestCertSeenView cert = do
+      roundStartSlot <- perasRoundStart (getPerasCertRound cert)
+      arrivalSlot <- perasCertArrivalSlot cert
+      pure $
+        LatestCertSeenView
+          { lcsCert = cert
+          , lcsArrivalSlot = arrivalSlot
+          , lcsRoundStartSlot = roundStartSlot
+          , lcsCandidateBlockExtendsCert = candidateBlockExtendsCert cert
+          }
+
+    mkLatestCertOnChainView cert =
+      pure $
+        LatestCertOnChainView
+          { lcocCert = cert
+          }
+
+    -- Does the candidate block extend the one boosted by a certificate?
+    --
+    -- This can be trivially tested by checking whether the certificate is
+    -- within the bounds of the chain prefix leading to the candidate block.
+    -- Conversely, the case of a certificate pointing to a block that's too new
+    -- to be voted for is also covered by this logic, as it won't be part of the
+    -- 'chainAtCandidateBlock' fragment.
+    --
+    -- NOTE: in the case of an extremely old certificate boosting a block beyond
+    -- the immutable prefix, this could incorrectly return false even if the
+    -- voting candidate technically extends the certificate's boosted block.
+    -- This case /should not/ affect the voting rules: this value is only used
+    -- for the "happy path", and the incorrect value can only happen during a
+    -- cooldown period where the "happy path" rules are not relevant.
+    --
+    -- Nevertheless, we should fix this by enhancing our view of the latest
+    -- certificate seen to know whether it is older than the volatile suffix:
+    -- TODO: https://github.com/tweag/cardano-peras/issues/186
+    candidateBlockExtendsCert cert =
+      AF.withinFragmentBounds
+        (castPoint (getPerasCertBoostedBlock cert))
+        chainAtCandidateBlock
+
+-- | Reason for voting being disallowed
```

Comment:
I don't think the newtypes around `Evidence` are very useful here.

I would either make a simple sum type `data VotingStatusWithEvidence = Vote (Evidence True PerasVotingRule) | NoVote (Evidence False PerasVotingRule)`,

or define `type VotingStatusWithEvidence b = Evidence b PerasVotingRule` and return `Either (VotingStatusWithEvidence 'True) (VotingStatusWithEvidence 'False)`.

But I think the first option is better

Reply:
Replace the newtypes with a GADT `PerasVotingRulesDecision`:

```haskell
data PerasVotingRulesDecision
  = Vote (Evidence True PerasVotingRule)
  | NoVote (Evidence False PerasVotingRule)
  deriving Show
```